// @ngx-formly/material/esm5/datepicker/datepicker.type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewChild, TemplateRef } from '@angular/core';
import { FieldType } from '@ngx-formly/material/form-field';
import { MatInput } from '@angular/material/input';
import { MatDatepickerInput } from '@angular/material/datepicker';
import { FormlyConfig, ɵwrapProperty as wrapProperty } from '@ngx-formly/core';
var FormlyFieldDatepicker = /** @class */ (function (_super) {
    tslib_1.__extends(FormlyFieldDatepicker, _super);
    function FormlyFieldDatepicker(config) {
        var _this = _super.call(this) || this;
        _this.config = config;
        _this.defaultOptions = {
            templateOptions: {
                datepickerOptions: {
                    startView: 'month',
                    datepickerTogglePosition: 'suffix',
                    dateInput: (/**
                     * @return {?}
                     */
                    function () { }),
                    dateChange: (/**
                     * @return {?}
                     */
                    function () { }),
                    monthSelected: (/**
                     * @return {?}
                     */
                    function () { }),
                    yearSelected: (/**
                     * @return {?}
                     */
                    function () { }),
                },
            },
        };
        return _this;
    }
    /**
     * @return {?}
     */
    FormlyFieldDatepicker.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        _super.prototype.ngAfterViewInit.call(this);
        // temporary fix for https://github.com/angular/components/issues/16761
        if (this.config.getValidatorMessage('matDatepickerParse')) {
            wrapProperty(this.field.formControl, 'errors', (/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var currentValue = _a.currentValue;
                if (currentValue && currentValue.required && currentValue.matDatepickerParse) {
                    /** @type {?} */
                    var errors = Object.keys(currentValue)
                        .sort((/**
                     * @param {?} prop
                     * @return {?}
                     */
                    function (prop) { return prop === 'matDatepickerParse' ? -1 : 0; }))
                        .reduce((/**
                     * @param {?} errors
                     * @param {?} prop
                     * @return {?}
                     */
                    function (errors, prop) {
                        var _a;
                        return (tslib_1.__assign({}, errors, (_a = {}, _a[prop] = currentValue[prop], _a)));
                    }), {});
                    ((/** @type {?} */ (_this.field.formControl)))['___$errors'] = errors;
                }
            }));
        }
        // temporary fix for https://github.com/angular/material2/issues/6728
        ((/** @type {?} */ (this.datepickerInput)))._formField = this.formField;
        setTimeout((/**
         * @return {?}
         */
        function () {
            switch (_this.to.datepickerOptions.datepickerTogglePosition) {
                case 'suffix':
                    _this.to._matSuffix = _this.datepickerToggle;
                    break;
                case 'prefix':
                    _this.to._matPrefix = _this.datepickerToggle;
                    break;
            }
            ((/** @type {?} */ (_this.options)))._markForCheck(_this.field);
        }));
    };
    /**
     * @return {?}
     */
    FormlyFieldDatepicker.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.fieldErrorsObserver && this.fieldErrorsObserver();
    };
    FormlyFieldDatepicker.decorators = [
        { type: Component, args: [{
                    selector: 'formly-field-mat-datepicker',
                    template: "\n    <input matInput\n      [id]=\"id\"\n      [errorStateMatcher]=\"errorStateMatcher\"\n      [formControl]=\"formControl\"\n      [matDatepicker]=\"picker\"\n      [matDatepickerFilter]=\"to.datepickerOptions.filter\"\n      [max]=\"to.datepickerOptions.max\"\n      [min]=\"to.datepickerOptions.min\"\n      [formlyAttributes]=\"field\"\n      [placeholder]=\"to.placeholder\"\n      [tabindex]=\"to.tabindex\"\n      [readonly]=\"to.readonly\"\n      [required]=\"to.required\"\n      (dateInput)=\"to.datepickerOptions.dateInput(field, $event)\"\n      (dateChange)=\"to.datepickerOptions.dateChange(field, $event)\">\n    <ng-template #datepickerToggle>\n      <mat-datepicker-toggle [disabled]=\"to.disabled\" [for]=\"picker\"></mat-datepicker-toggle>\n    </ng-template>\n    <mat-datepicker #picker\n      [color]=\"to.color\"\n      [dateClass]=\"to.datepickerOptions.dateClass\"\n      [disabled]=\"to.datepickerOptions.disabled\"\n      [opened]=\"to.datepickerOptions.opened\"\n      [panelClass]=\"to.datepickerOptions.panelClass\"\n      [startAt]=\"to.datepickerOptions.startAt\"\n      [startView]=\"to.datepickerOptions.startView\"\n      [touchUi]=\"to.datepickerOptions.touchUi\"\n      (monthSelected)=\"to.datepickerOptions.monthSelected(field, $event, picker)\"\n      (yearSelected)=\"to.datepickerOptions.yearSelected(field, $event, picker)\"\n    >\n    </mat-datepicker>\n  "
                }] }
    ];
    /** @nocollapse */
    FormlyFieldDatepicker.ctorParameters = function () { return [
        { type: FormlyConfig }
    ]; };
    FormlyFieldDatepicker.propDecorators = {
        formFieldControl: [{ type: ViewChild, args: [MatInput, (/** @type {?} */ ({ static: true })),] }],
        datepickerInput: [{ type: ViewChild, args: [MatDatepickerInput, (/** @type {?} */ ({ static: true })),] }],
        datepickerToggle: [{ type: ViewChild, args: ['datepickerToggle',] }]
    };
    return FormlyFieldDatepicker;
}(FieldType));
export { FormlyFieldDatepicker };
if (false) {
    /** @type {?} */
    FormlyFieldDatepicker.prototype.formFieldControl;
    /** @type {?} */
    FormlyFieldDatepicker.prototype.datepickerInput;
    /** @type {?} */
    FormlyFieldDatepicker.prototype.datepickerToggle;
    /** @type {?} */
    FormlyFieldDatepicker.prototype.defaultOptions;
    /**
     * @type {?}
     * @private
     */
    FormlyFieldDatepicker.prototype.fieldErrorsObserver;
    /**
     * @type {?}
     * @private
     */
    FormlyFieldDatepicker.prototype.config;
}
