// @ngx-formly/material/bundles/ngx-formly-material-datepicker.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/forms'), require('@angular/core'), require('@ngx-formly/material/form-field'), require('@angular/material/input'), require('@angular/material/datepicker'), require('@ngx-formly/core')) :
    typeof define === 'function' && define.amd ? define('@ngx-formly/material/datepicker', ['exports', '@angular/common', '@angular/forms', '@angular/core', '@ngx-formly/material/form-field', '@angular/material/input', '@angular/material/datepicker', '@ngx-formly/core'], factory) :
    (factory((global['ngx-formly'] = global['ngx-formly'] || {}, global['ngx-formly'].material = global['ngx-formly'].material || {}, global['ngx-formly'].material.datepicker = {}),global.ng.common,global.ng.forms,global.ng.core,global['ngx-formly'].material['form-field'],global.ng.material.input,global.ng.material.datepicker,global.core$1));
}(this, (function (exports,common,forms,core,formField,input,datepicker,core$1) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormlyFieldDatepicker = /** @class */ (function (_super) {
        __extends(FormlyFieldDatepicker, _super);
        function FormlyFieldDatepicker(config) {
            var _this = _super.call(this) || this;
            _this.config = config;
            _this.defaultOptions = {
                templateOptions: {
                    datepickerOptions: {
                        startView: 'month',
                        datepickerTogglePosition: 'suffix',
                        dateInput: ( /**
                         * @return {?}
                         */function () { }),
                        dateChange: ( /**
                         * @return {?}
                         */function () { }),
                        monthSelected: ( /**
                         * @return {?}
                         */function () { }),
                        yearSelected: ( /**
                         * @return {?}
                         */function () { }),
                    },
                },
            };
            return _this;
        }
        /**
         * @return {?}
         */
        FormlyFieldDatepicker.prototype.ngAfterViewInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                _super.prototype.ngAfterViewInit.call(this);
                // temporary fix for https://github.com/angular/components/issues/16761
                if (this.config.getValidatorMessage('matDatepickerParse')) {
                    core$1.ɵwrapProperty(this.field.formControl, 'errors', ( /**
                     * @param {?} __0
                     * @return {?}
                     */function (_a) {
                        var currentValue = _a.currentValue;
                        if (currentValue && currentValue.required && currentValue.matDatepickerParse) {
                            /** @type {?} */
                            var errors = Object.keys(currentValue)
                                .sort(( /**
                         * @param {?} prop
                         * @return {?}
                         */function (prop) { return prop === 'matDatepickerParse' ? -1 : 0; }))
                                .reduce(( /**
                         * @param {?} errors
                         * @param {?} prop
                         * @return {?}
                         */function (errors, prop) {
                                var _a;
                                return (__assign({}, errors, (_a = {}, _a[prop] = currentValue[prop], _a)));
                            }), {});
                            (( /** @type {?} */(_this.field.formControl)))['___$errors'] = errors;
                        }
                    }));
                }
                // temporary fix for https://github.com/angular/material2/issues/6728
                (( /** @type {?} */(this.datepickerInput)))._formField = this.formField;
                setTimeout(( /**
                 * @return {?}
                 */function () {
                    switch (_this.to.datepickerOptions.datepickerTogglePosition) {
                        case 'suffix':
                            _this.to._matSuffix = _this.datepickerToggle;
                            break;
                        case 'prefix':
                            _this.to._matPrefix = _this.datepickerToggle;
                            break;
                    }
                    (( /** @type {?} */(_this.options)))._markForCheck(_this.field);
                }));
            };
        /**
         * @return {?}
         */
        FormlyFieldDatepicker.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                this.fieldErrorsObserver && this.fieldErrorsObserver();
            };
        FormlyFieldDatepicker.decorators = [
            { type: core.Component, args: [{
                        selector: 'formly-field-mat-datepicker',
                        template: "\n    <input matInput\n      [id]=\"id\"\n      [errorStateMatcher]=\"errorStateMatcher\"\n      [formControl]=\"formControl\"\n      [matDatepicker]=\"picker\"\n      [matDatepickerFilter]=\"to.datepickerOptions.filter\"\n      [max]=\"to.datepickerOptions.max\"\n      [min]=\"to.datepickerOptions.min\"\n      [formlyAttributes]=\"field\"\n      [placeholder]=\"to.placeholder\"\n      [tabindex]=\"to.tabindex\"\n      [readonly]=\"to.readonly\"\n      [required]=\"to.required\"\n      (dateInput)=\"to.datepickerOptions.dateInput(field, $event)\"\n      (dateChange)=\"to.datepickerOptions.dateChange(field, $event)\">\n    <ng-template #datepickerToggle>\n      <mat-datepicker-toggle [disabled]=\"to.disabled\" [for]=\"picker\"></mat-datepicker-toggle>\n    </ng-template>\n    <mat-datepicker #picker\n      [color]=\"to.color\"\n      [dateClass]=\"to.datepickerOptions.dateClass\"\n      [disabled]=\"to.datepickerOptions.disabled\"\n      [opened]=\"to.datepickerOptions.opened\"\n      [panelClass]=\"to.datepickerOptions.panelClass\"\n      [startAt]=\"to.datepickerOptions.startAt\"\n      [startView]=\"to.datepickerOptions.startView\"\n      [touchUi]=\"to.datepickerOptions.touchUi\"\n      (monthSelected)=\"to.datepickerOptions.monthSelected(field, $event, picker)\"\n      (yearSelected)=\"to.datepickerOptions.yearSelected(field, $event, picker)\"\n    >\n    </mat-datepicker>\n  "
                    }] }
        ];
        /** @nocollapse */
        FormlyFieldDatepicker.ctorParameters = function () {
            return [
                { type: core$1.FormlyConfig }
            ];
        };
        FormlyFieldDatepicker.propDecorators = {
            formFieldControl: [{ type: core.ViewChild, args: [input.MatInput, ( /** @type {?} */({ static: true })),] }],
            datepickerInput: [{ type: core.ViewChild, args: [datepicker.MatDatepickerInput, ( /** @type {?} */({ static: true })),] }],
            datepickerToggle: [{ type: core.ViewChild, args: ['datepickerToggle',] }]
        };
        return FormlyFieldDatepicker;
    }(formField.FieldType));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormlyMatDatepickerModule = /** @class */ (function () {
        function FormlyMatDatepickerModule() {
        }
        FormlyMatDatepickerModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [FormlyFieldDatepicker],
                        imports: [
                            common.CommonModule,
                            forms.ReactiveFormsModule,
                            input.MatInputModule,
                            datepicker.MatDatepickerModule,
                            formField.FormlyMatFormFieldModule,
                            core$1.FormlyModule.forChild({
                                types: [{
                                        name: 'datepicker',
                                        component: FormlyFieldDatepicker,
                                        wrappers: ['form-field'],
                                    }],
                            }),
                        ],
                    },] }
        ];
        return FormlyMatDatepickerModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.FormlyMatDatepickerModule = FormlyMatDatepickerModule;
    exports.FormlyFieldDatepicker = FormlyFieldDatepicker;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
