// @ngx-lite/input-star-rating/esm2015/ngx-input-star-rating.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxInputStarRatingComponent } from './ngx-input-star-rating.component';
import * as i0 from "@angular/core";
export class NgxInputStarRatingModule {
}
NgxInputStarRatingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInputStarRatingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxInputStarRatingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInputStarRatingModule, declarations: [NgxInputStarRatingComponent], imports: [CommonModule], exports: [NgxInputStarRatingComponent] });
NgxInputStarRatingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInputStarRatingModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInputStarRatingModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [NgxInputStarRatingComponent],
                    exports: [NgxInputStarRatingComponent],
                }]
        }] });
