// ngx-page-scroll/esm2020/lib/ngx-page-scroll.module.mjs
import { NgModule } from '@angular/core';
import { NgxPageScrollCoreModule } from 'ngx-page-scroll-core';
import { NgxPageScrollDirective } from './ngx-page-scroll.directive';
import * as i0 from "@angular/core";
export class NgxPageScrollModule {
}
NgxPageScrollModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: NgxPageScrollModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxPageScrollModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: NgxPageScrollModule, declarations: [NgxPageScrollDirective], imports: [NgxPageScrollCoreModule], exports: [NgxPageScrollDirective] });
NgxPageScrollModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: NgxPageScrollModule, imports: [[
            NgxPageScrollCoreModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.2", ngImport: i0, type: NgxPageScrollModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        NgxPageScrollDirective,
                    ],
                    imports: [
                        NgxPageScrollCoreModule,
                    ],
                    exports: [
                        NgxPageScrollDirective,
                    ],
                }]
        }] });
