// @ngxs-labs/immer-adapter/esm2015/lib/core/immer-adapter/common/immutable-state-context.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { createDraft, finishDraft } from 'immer';
/**
 * @template T
 */
export class ImmutableStateContext {
    /**
     * @param {?} ctx
     */
    constructor(ctx) {
        this.ctx = ctx;
        this.frozenState = null;
        ImmutableStateContext.autobindStateContext(this);
    }
    /**
     * @private
     * @param {?} context
     * @return {?}
     */
    static autobindStateContext(context) {
        for (const prop of Object.getOwnPropertyNames(Object.getPrototypeOf(context))) {
            if (prop === 'constructor' || typeof context[prop] !== 'function') {
                continue;
            }
            context[prop] = context[prop].bind(context);
        }
    }
    /**
     * @return {?}
     */
    getState() {
        this.frozenState = (/** @type {?} */ (createDraft(this.ctx.getState())));
        return this.frozenState;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    setState(val) {
        /** @type {?} */
        let state;
        if (typeof val === 'function') {
            /** @type {?} */
            let newState;
            /** @type {?} */
            const oldState = (/** @type {?} */ (createDraft(this.ctx.getState())));
            /** @type {?} */
            const operator = (/** @type {?} */ (val));
            /** @type {?} */
            const mutatedOldState = operator(oldState);
            if (this.frozenState === mutatedOldState) {
                newState = finishDraft(this.frozenState);
                finishDraft(oldState);
            }
            else {
                /** @type {?} */
                const mutateOutsideOperator = oldState !== mutatedOldState;
                if (mutateOutsideOperator) {
                    newState = mutatedOldState;
                    finishDraft(oldState);
                }
                else {
                    newState = finishDraft(mutatedOldState);
                }
            }
            state = newState;
        }
        else {
            state = finishDraft(val);
        }
        this.frozenState = null;
        return this.ctx.setState(state);
    }
    /**
     * @param {?} val
     * @return {?}
     */
    patchState(val) {
        return this.ctx.patchState((/** @type {?} */ (finishDraft(val))));
    }
    /**
     * @param {?} actions
     * @return {?}
     */
    dispatch(actions) {
        return this.ctx.dispatch(actions);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    ImmutableStateContext.prototype.frozenState;
    /**
     * @type {?}
     * @private
     */
    ImmutableStateContext.prototype.ctx;
}
