// @ngxs-labs/immer-adapter/esm5/lib/core/immer-adapter/common/immutable-state-context.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { createDraft, finishDraft } from 'immer';
/**
 * @template T
 */
var /**
 * @template T
 */
ImmutableStateContext = /** @class */ (function () {
    function ImmutableStateContext(ctx) {
        this.ctx = ctx;
        this.frozenState = null;
        ImmutableStateContext.autobindStateContext(this);
    }
    /**
     * @private
     * @param {?} context
     * @return {?}
     */
    ImmutableStateContext.autobindStateContext = /**
     * @private
     * @param {?} context
     * @return {?}
     */
    function (context) {
        var e_1, _a;
        try {
            for (var _b = tslib_1.__values(Object.getOwnPropertyNames(Object.getPrototypeOf(context))), _c = _b.next(); !_c.done; _c = _b.next()) {
                var prop = _c.value;
                if (prop === 'constructor' || typeof context[prop] !== 'function') {
                    continue;
                }
                context[prop] = context[prop].bind(context);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    /**
     * @return {?}
     */
    ImmutableStateContext.prototype.getState = /**
     * @return {?}
     */
    function () {
        this.frozenState = (/** @type {?} */ (createDraft(this.ctx.getState())));
        return this.frozenState;
    };
    /**
     * @param {?} val
     * @return {?}
     */
    ImmutableStateContext.prototype.setState = /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        /** @type {?} */
        var state;
        if (typeof val === 'function') {
            /** @type {?} */
            var newState = void 0;
            /** @type {?} */
            var oldState = (/** @type {?} */ (createDraft(this.ctx.getState())));
            /** @type {?} */
            var operator = (/** @type {?} */ (val));
            /** @type {?} */
            var mutatedOldState = operator(oldState);
            if (this.frozenState === mutatedOldState) {
                newState = finishDraft(this.frozenState);
                finishDraft(oldState);
            }
            else {
                /** @type {?} */
                var mutateOutsideOperator = oldState !== mutatedOldState;
                if (mutateOutsideOperator) {
                    newState = mutatedOldState;
                    finishDraft(oldState);
                }
                else {
                    newState = finishDraft(mutatedOldState);
                }
            }
            state = newState;
        }
        else {
            state = finishDraft(val);
        }
        this.frozenState = null;
        return this.ctx.setState(state);
    };
    /**
     * @param {?} val
     * @return {?}
     */
    ImmutableStateContext.prototype.patchState = /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        return this.ctx.patchState((/** @type {?} */ (finishDraft(val))));
    };
    /**
     * @param {?} actions
     * @return {?}
     */
    ImmutableStateContext.prototype.dispatch = /**
     * @param {?} actions
     * @return {?}
     */
    function (actions) {
        return this.ctx.dispatch(actions);
    };
    return ImmutableStateContext;
}());
/**
 * @template T
 */
export { ImmutableStateContext };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ImmutableStateContext.prototype.frozenState;
    /**
     * @type {?}
     * @private
     */
    ImmutableStateContext.prototype.ctx;
}
