// @ngx-lite/input-tag/esm2015/ngx-input-tag.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxInputTagComponent, formatter as tagFormatter, } from './ngx-input-tag.component';
import { NGX_INPUT_TAG_TAG_FORMATTER } from './ngx-input-tag.di-tokens';
import * as i0 from "@angular/core";
export class NgxInputTagModule {
    static forRoot(config = { tagFormatter }) {
        return {
            ngModule: NgxInputTagModule,
            providers: [
                {
                    provide: NGX_INPUT_TAG_TAG_FORMATTER,
                    useValue: config.tagFormatter,
                },
            ],
        };
    }
}
NgxInputTagModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInputTagModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxInputTagModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInputTagModule, declarations: [NgxInputTagComponent], imports: [CommonModule], exports: [NgxInputTagComponent] });
NgxInputTagModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInputTagModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInputTagModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [NgxInputTagComponent],
                    exports: [NgxInputTagComponent],
                }]
        }] });
