// @ngx-lite/in-viewport/esm2015/ngx-in-viewport.directive.js
import { Directive, EventEmitter, Inject, Input, Output, PLATFORM_ID, } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { fromEvent } from 'rxjs';
import { debounceTime, merge } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class NgxInViewportDirective {
    constructor(platformId, elementRef) {
        this.platformId = platformId;
        this.elementRef = elementRef;
        this.offset = 0;
        this.inViewport = new EventEmitter();
    }
    ngOnInit() {
        if (isPlatformBrowser(this.platformId)) {
            this.subscription = fromEvent(window, 'scroll')
                .pipe(merge(fromEvent(window, 'resize')), debounceTime(100))
                .subscribe(() => this.check());
        }
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
    check() {
        const event = {
            target: this.elementRef.nativeElement,
            value: document.body.contains(this.elementRef.nativeElement) &&
                this.elementRef.nativeElement.getBoundingClientRect().top <=
                    window.innerHeight + this.offset,
        };
        this.inViewport.emit(event);
    }
}
NgxInViewportDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInViewportDirective, deps: [{ token: PLATFORM_ID }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
NgxInViewportDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.2", type: NgxInViewportDirective, selector: "[ngxInViewport]", inputs: { offset: "offset" }, outputs: { inViewport: "inViewport" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInViewportDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngxInViewport]',
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [PLATFORM_ID]
                }] }, { type: i0.ElementRef }]; }, propDecorators: { offset: [{
                type: Input
            }], inViewport: [{
                type: Output
            }] } });
