// @ngx-lite/in-viewport/esm2015/ngx-in-viewport.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxInViewportDirective } from './ngx-in-viewport.directive';
import * as i0 from "@angular/core";
export class NgxInViewportModule {
}
NgxInViewportModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInViewportModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxInViewportModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInViewportModule, declarations: [NgxInViewportDirective], imports: [CommonModule], exports: [NgxInViewportDirective] });
NgxInViewportModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInViewportModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0, type: NgxInViewportModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [NgxInViewportDirective],
                    exports: [NgxInViewportDirective],
                }]
        }] });
