// @ngx-lite/input-star-rating/bundles/ngx-lite-input-star-rating.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@ngx-lite/input-star-rating', ['exports', '@angular/core', '@angular/forms', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['ngx-lite'] = global['ngx-lite'] || {}, global['ngx-lite']['input-star-rating'] = {}), global.ng.core, global.ng.forms, global.ng.common));
}(this, (function (exports, i0, forms, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () {
                            return e[k];
                        }
                    });
                }
            });
        }
        n['default'] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var starIcons = {
        outline: 'outline',
        full: 'full',
        half: 'half',
    };
    var NgxInputStarRatingComponent = /** @class */ (function () {
        function NgxInputStarRatingComponent() {
            this._value = 0;
            this.count = 5;
            this.disabled = false;
            this.starIcons = starIcons;
            this.buttons = [];
            // tslint:disable-next-line:no-empty
            this.onChange = function (_value) { };
            // tslint:disable-next-line:no-empty
            this.onTouched = function () { };
        }
        Object.defineProperty(NgxInputStarRatingComponent.prototype, "value", {
            get: function () {
                return this._value;
            },
            set: function (val) {
                this._value = val;
                this.onChange(val);
                this.valueUpdate(this._value);
                this.onTouched();
            },
            enumerable: false,
            configurable: true
        });
        NgxInputStarRatingComponent.prototype.registerOnChange = function (fn) {
            this.onChange = fn;
        };
        NgxInputStarRatingComponent.prototype.registerOnTouched = function (fn) {
            this.onTouched = fn;
        };
        NgxInputStarRatingComponent.prototype.writeValue = function (value) {
            if (value !== this.value) {
                this.value = value;
            }
        };
        NgxInputStarRatingComponent.prototype.ngOnInit = function () {
            for (var i = 0; i < this.count; i++) {
                this.buttons.push({ active: false, icon: starIcons.outline });
            }
            this.valueUpdate(this.value);
        };
        NgxInputStarRatingComponent.prototype.ngOnChanges = function (changes) {
            if (changes.value && changes.value.currentValue) {
                this.valueUpdate(changes.value.currentValue);
            }
        };
        NgxInputStarRatingComponent.prototype.rate = function (index) {
            this.buttons.forEach(function (btn, i) {
                if (i <= index) {
                    btn.icon = starIcons.full;
                }
                else {
                    btn.icon = starIcons.outline;
                }
            });
            var total = this.buttons.reduce(function (prev, next) {
                if (next.icon === starIcons.full) {
                    return prev + 1;
                }
                else {
                    return prev;
                }
            }, 0);
            this.value = total;
        };
        NgxInputStarRatingComponent.prototype.mouseEnter = function (index) {
            this.buttons.forEach(function (btn, i) {
                if (i <= index) {
                    btn.active = true;
                }
                else {
                    btn.active = false;
                }
            });
        };
        NgxInputStarRatingComponent.prototype.mouseLeave = function () {
            this.buttons.forEach(function (i) { return (i.active = false); });
        };
        NgxInputStarRatingComponent.prototype.valueUpdate = function (value) {
            this.buttons.forEach(function (btn, i) {
                var starValue = i + 1;
                if (value >= starValue) {
                    btn.icon = starIcons.full;
                }
                if (value >= starValue - 0.5 && value < starValue) {
                    btn.icon = starIcons.half;
                }
            });
        };
        return NgxInputStarRatingComponent;
    }());
    NgxInputStarRatingComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0__namespace, type: NgxInputStarRatingComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    NgxInputStarRatingComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.2", type: NgxInputStarRatingComponent, selector: "ngx-input-star-rating", inputs: { _value: ["value", "_value"], count: "count", disabled: "disabled" }, providers: [
            {
                provide: forms.NG_VALUE_ACCESSOR,
                useExisting: i0.forwardRef(function () { return NgxInputStarRatingComponent; }),
                multi: true,
            },
        ], usesOnChanges: true, ngImport: i0__namespace, template: "<div class=\"ngx-input-rating\" (mouseleave)=\"mouseLeave()\">\n  <button type=\"button\" [disabled]=\"disabled\" *ngFor=\"let btn of buttons; let i = index\" class=\"ngx-input-rating__btn\" (click)=\"rate(i)\" [attr.aria-label]=\"'rate ' + (i + 1) + ' of ' + buttons.length\"\n    (mouseenter)=\"mouseEnter(i)\" class=\"ngx-input-star-rating__btn\">\n    <span *ngIf=\"btn.active || btn.icon === starIcons.full\" class=\"ngx-input-star-rating__full\">\n      <ng-container *ngTemplateOutlet=\"star\"></ng-container>\n    </span>\n    <ng-container *ngIf=\"!btn.active\">\n      <span *ngIf=\"btn.icon === starIcons.half\" class=\"ngx-input-star-rating__half\">\n        <ng-container *ngTemplateOutlet=\"star\"></ng-container>\n        <svg height=\"25\" width=\"10\" class=\"ngx-input-star-rating__star-half\">\n          <polygon points=\"9.9, 1.1, 3.3, 21.78, 19.8, 8.58, 0, 8.58, 16.5, 21.78\" style=\"fill-rule:nonzero;\"/>\n        </svg>\n      </span>\n      <span *ngIf=\"btn.icon !== starIcons.full && btn.icon !== starIcons.half\" class=\"empty\">\n        <ng-container *ngTemplateOutlet=\"star\"></ng-container>\n      </span>\n    </ng-container>\n  </button>\n</div>\n\n<ng-template #star>\n  <svg height=\"25\" width=\"23\" class=\"ngx-input-star-rating__star\">\n    <polygon points=\"9.9, 1.1, 3.3, 21.78, 19.8, 8.58, 0, 8.58, 16.5, 21.78\" style=\"fill-rule:nonzero;\"/>\n  </svg>\n</ng-template>", styles: ["*,:after,:before{box-sizing:border-box}ngx-input-star-rating{display:inline-block;--color-star-default:#ccc;--color-star-highlight:#3d3d3d}.ngx-input-star-rating__star{fill:#ccc;fill:var(--color-star-default)}.ngx-input-star-rating__btn{background-color:initial;border:0;cursor:pointer;position:relative;width:30px;height:25px}.ngx-input-star-rating__btn svg{position:absolute;top:0;left:5px}.ngx-input-star-rating__full .ngx-input-star-rating__star,.ngx-input-star-rating__half .ngx-input-star-rating__star-half{fill:#3d3d3d;fill:var(--color-star-highlight)}"], directives: [{ type: i1__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i1__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i1__namespace.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0__namespace, type: NgxInputStarRatingComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'ngx-input-star-rating',
                        templateUrl: './ngx-input-star-rating.component.html',
                        styleUrls: ['./ngx-input-star-rating.component.scss'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            {
                                provide: forms.NG_VALUE_ACCESSOR,
                                useExisting: i0.forwardRef(function () { return NgxInputStarRatingComponent; }),
                                multi: true,
                            },
                        ],
                        encapsulation: i0.ViewEncapsulation.None,
                    }]
            }], propDecorators: { _value: [{
                    type: i0.Input,
                    args: ['value']
                }], count: [{
                    type: i0.Input
                }], disabled: [{
                    type: i0.Input
                }] } });

    var NgxInputStarRatingModule = /** @class */ (function () {
        function NgxInputStarRatingModule() {
        }
        return NgxInputStarRatingModule;
    }());
    NgxInputStarRatingModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0__namespace, type: NgxInputStarRatingModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    NgxInputStarRatingModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0__namespace, type: NgxInputStarRatingModule, declarations: [NgxInputStarRatingComponent], imports: [i1.CommonModule], exports: [NgxInputStarRatingComponent] });
    NgxInputStarRatingModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0__namespace, type: NgxInputStarRatingModule, imports: [[i1.CommonModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.2", ngImport: i0__namespace, type: NgxInputStarRatingModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i1.CommonModule],
                        declarations: [NgxInputStarRatingComponent],
                        exports: [NgxInputStarRatingComponent],
                    }]
            }] });

    /*
     * Public API Surface of ngx-input-star-rating
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.NgxInputStarRatingComponent = NgxInputStarRatingComponent;
    exports.NgxInputStarRatingModule = NgxInputStarRatingModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
